tokio/src/io/async_fd.rs

```diff
@@ -0,0 +1,269 @@
+/// when the file descriptor transitions from not-ready to ready. Tokio
+/// internally tracks when it has received a ready notification, and when
+/// readiness checking functions like [`readable`] and [`writable`] are called,
+/// if the readiness flag is set, these async functions will complete
+/// immediately.
+///
+/// This however does mean that it is critical to ensure that this ready flag is
+/// cleared when (and only when) the file descriptor ceases to be ready. The
+/// [`ReadyGuard`] returned from readiness checking functions serves this
+/// function; after calling a readiness-checking async function, you must use
+/// this [`ReadyGuard`] to signal to tokio whether the file descriptor is no
+/// longer in a ready state.
+///
+/// ## Converting to a poll-based API
+///
+/// In some cases it may be desirable to use `AsyncFd` from APIs similar to
+/// [`TcpStream::poll_read_ready`]. One can do so by allocating a pinned future
+/// to perform the poll:
+///
+/// ```
+/// use tokio::io::{ReadyGuard, AsyncFd};
+///
+/// use std::future::Future;
+/// use std::sync::Arc;
+/// use std::pin::Pin;
+/// use std::task::{Context, Poll};
+///
+/// use futures::ready;
+///
+/// struct MyIoStruct {
+///   async_fd: Arc<AsyncFd>,
+///   poller: Pin<Box<dyn Future<Output=()>>>
+/// }
+///
+/// impl MyIoStruct {
+///    fn poll_read_ready(&mut self, cx: &mut Context<'_>) -> Poll<std::io::Result<ReadyGuard<'_>>> {
+///        let mut result = Poll::Pending;
+///        while result.is_pending() {
+///            // Poll the saved future; if it's not ready, our context waker will be saved in the
+///            // future and we can return.
+///            ready!(self.poller.as_mut().poll(cx));
+///
+///            // Reset the poller future, since we consumed it.
+///            let arc = self.async_fd.clone();
+///            self.poller = Box::pin(async move {
+///                let _ = arc.readable().await.map(|mut guard| guard.retain_ready());
+///            });
+///
+///            // Because we need to bind the ReadyGuard to the lifetime of self, we have to re-poll here.
+///            // It's possible that we might race with another thread clearing the ready state, so deal
+///            // with that as well.
+///            let fut = self.async_fd.readable();
+///            tokio::pin!(fut);
+///            result = fut.as_mut().poll(cx);
+///        }
+///
+///        result
+///    }
+/// }
+/// ```
+///
+/// [`readable`]: method@Self::readable
+/// [`writable`]: method@Self::writable
+/// [`ReadyGuard`]: struct@self::ReadyGuard
+/// [`TcpStream::poll_read_ready`]: struct@crate::net::TcpStream
+pub struct AsyncFd {
+    handle: Handle,
+    fd: RawFd,
+    shared: slab::Ref<ScheduledIo>,
+}
+
+impl std::fmt::Debug for AsyncFd {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("AsyncFd").field("fd", &self.fd).finish()
+    }
+}
+
+unsafe impl Send for AsyncFd {}
```

Comment:
Do we know which field *isn't* `Send`/`Sync` of `AsyncFd`? Not a blocker, but would just like to know what *isn't* so we know how to be safe in this type.

Reply:
I would guess `slab::Ref`. That probably should be changed...

Reply:
turns out this isn't needed (and is removed in the latest revision)